### DISCUSSION TITLE
Need for testimonials about using VS Code & Z Open Editor

### DISCUSSION THREAD

**Discussion by FALLAI-Denis** [IBM/zopeneditor-about#139] [Q&A]
## Testimonials

Hi, 

This may not be the purpose of the GitHub site to make this request, but I have not identified other ways to do so.
If my request is a problem for the administrators of the site, I authorize them to mean it in response to this message, and to close this issue.

We are in the process of defining our new COBOL development workshop, which should replace our current solution based on a COBOL code generator (IBM VisualAge Pacbase, in use for us since 1988, still active, but no longer maintained).

We are convinced of the qualities of the VS Code product and the Z Open Editor extension, and that we are making the right choice by choosing these solutions. However, the highest authorities of our company ask us to justify our choices and to validate them with major market compagnies.

These solutions are recent and not widely used, at least in France, and we have not succeeded at this time in finding users testimonials who have already implemented these solutions and already use them on a daily basis.

I am therefore asking the users of this GitHub site who could give us their testimony in the use of the VS Code and Z Open Editor solutions in a more advanced context than ours:
* what feeling on the part of the users (developers),
* what productivity obtained by developers compared to another earlier solution,
* what difficulties may have been encountered in setting up the solution,
* what additional developments were necessary to finalize the solution,
* what are the potential risks, on the products, and on their implementation, on the impact on users,
* etc.

Some information about our context:
* between 600 and 700 developers, 50% in service with a significant turnover,
* between 300 and 400 active developers per day on COBOL coding activities,
* around 30,000 COBOL programs active in production, (note: without any COPYBOOK because it does not make sense with the use of a code generator ... we have work to do on adapting this legacy to maintain it in a traditional way),
* around 1,800 applications, each application to be managed in its Git repository,
* COBOL (V6) /  Endevor / CICS / Db2 / MQ technical background

Answers can be provided on this GitHub site, if the administrators authorize it, or by personal messaging.

Thanking in advance the people who will be able to answer my request.

**Comment by phaumer** [comment]
Thanks. We would definitely be interested in responses to these questions as well.

**Comment by bearou** [comment]
Hi; (better late than never :))
We're on the brink of making VSCode and Z Open Editor an official alternative IDE to developing COBOL programs. An alternative to a legacy homegrown ISPF-based solution, having experimented with Zowe the last couple of years... On top of that we're also introducing Git (Bitbucket) as a source repo, but this is not forced yet.

We are similar convinced that VSCode and Z Open Editor will be preferred by younger developers having their Java-programs and Jenkins build/deploy pipelines in the same toolset.

Some information about our context:

Approx 100 active developers per day on COBOL coding activities,
around 110,000 COBOL programs active in production, including copybooks (and some machine converted stuff too).
COBOL (V6) / selfmade IDE and source repo (ISPF) / Batch / CICS / Db2 / MQ
As a potential risk, we're currently facing performance difficulties in the wast amout of COBOL modules and libraries that we have. 1,4 seconds is the current average time for Z Open Editor to fetch a copybook when resolving a COBOL program... 1,4 seconds times 231 (maximum copybooks) is simply too much... Hence, we're looking for performance tuning options, but are currently unable to find a good advices.

**Reply by phaumer** [reply]
> As a potential risk, we're currently facing performance difficulties in the wast amout of COBOL modules and libraries that we have. 1,4 seconds is the current average time for Z Open Editor to fetch a copybook when resolving a COBOL program... 1,4 seconds times 231 (maximum copybooks) is simply too much... Hence, we're looking for performance tuning options, but are currently unable to find a good advices.

Thanks for the great feedback. The editor currently downloads only up to three copybooks in parallel as we saw some people reporting issues with z/OSMF and many concurrent rest requests. We can better tweak RSE and less z/OSMF. However, we could expose these kind of behavior settings such as parallel downloads as user settings for people to play around with for tweaking. If you are interested, please file an issue.

There is also the `zowe.files.temporaryDownloadsFolder.cleanup` setting that Z Open Editor respects as well. We added that to allow editing in more than one window, but the side effect is also that downloaded copybooks survive the session, keeping your files locally in your temp folder (if your security policies allows that). That could also improve startup times with the side effect that some of your files are outdated as it does not check for updates.

**Comment by FALLAI-Denis** [comment]
Hi @bearou 

Thank you for your testimonial.

Concerning the access time to the Copybooks, which is linked to the access to the mainframe whether by z/OSMF or by RSE, there might be an alternative by using a management of the Copybooks in dedicated Git repositories, and by associating these Git Repositories with the main Git Repository of COBOL sources (or other), either by the mechanism of submodules, or by the mechanism of subtrees. In this case the Copybooks can become local to the Workspace.

This is something we need to experience as part of implementing the IBM Dependency Based Build solution.

[Support of Git submodules in Build process](https://github.com/IBM/dbb/issues/110 )
[All languages (COBOL first) - Search for COPYBOOKs / INCLUDEs in Git Repositories](https://github.com/IBM/zopeneditor-about/issues/233 )
